File: src/main/kotlin/dqualizer/dqexec/loadtest/mapper/k6/HttpMapper.java
package dqualizer.dqexec.loadtest.mapper.k6;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import dqualizer.dqexec.config.PathConfig;
import dqualizer.dqexec.exception.UnknownRequestTypeException;
import dqualizer.dqexec.util.SafeFileReader;
import dqualizer.dqlang.archive.k6adapter.dqlang.k6.request.Request;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.Iterator;
import java.util.Map;
import java.util.Optional;

/**
 * Maps the specified request to a k6 'default function()'
 */
@Component
@RequiredArgsConstructor
public class HttpMapper implements k6Mapper {

    private final PathConfig paths;
    private final SafeFileReader reader;

    @Override
    public String map(Request request) {
        StringBuilder httpBuilder = new StringBuilder();
        httpBuilder.append(this.exportFunctionScript());
        String path = request.getPath();
        String type = request.getType().toUpperCase();
        String method = switch (type) {
            case "GET" -> "get";
            case "POST" -> "post";
            case "PUT" -> "put";
            case "DELETE" -> "del";
            default -> throw new UnknownRequestTypeException(type);
        };

        //Code for choosing one random variable for every existing path variable
        Map<String, String> pathVariables = request.getPathVariables();
        Optional<String> maybeReference = pathVariables.values().stream().findFirst();
        if (maybeReference.isPresent()) {
            String referencePath = paths.getResourcePath() + maybeReference.get();
            String pathVariablesString = reader.readFile(referencePath);

            try {
                JsonNode node = new ObjectMapper().readTree(pathVariablesString);

                Iterator<String> variables = node.fieldNames();
                while (variables.hasNext()) {
                    String variable = variables.next();
                    String randomPathVariable = this.randomPathVaribleScript(variable);
                    httpBuilder.append(randomPathVariable);
                }

            } catch (JsonProcessingException e) {
                e.printStackTrace();
            }
        }

        //Code for using the request-parameter and payload inside the http method
        Map<String, String> payload = request.getPayload();
        Map<String, String> queryParams = request.getQueryParams();
        String extraParams = "";
        if (!payload.isEmpty() || !queryParams.isEmpty()) {
            if (!payload.isEmpty() && !queryParams.isEmpty()) {
                httpBuilder.append(this.randomQueryParamScript());
                httpBuilder.append(this.randomPayloadScript());
                extraParams = " + `?${urlSearchParams.toString()}`, JSON.stringify(payload)";
            } else if (!payload.isEmpty()) {
                httpBuilder.append(this.randomPayloadScript());
                extraParams = ", JSON.stringify(payload)";
            } else {
                httpBuilder.append(this.randomQueryParamScript());
                extraParams = " + `?${urlSearchParams.toString()}`";
            }
        }
        String httpRequest = String.format("%slet response = http.%s(baseURL + `%s`%s, params);%s",
                newLine, method, path, extraParams, newLine);
        httpBuilder.append(httpRequest);

        return httpBuilder.toString();
    }

    private String exportFunctionScript() {
        return String.format("%sexport default function() {%s",
                newLine, newLine);
    }

    private String randomPathVaribleScript(String variable) {
        return String.format("%slet %s = %s_array[Math.floor(Math.random() * %s_array.length)];%s",
                newLine, variable, variable, variable, newLine);
    }

    private String randomPayloadScript() {
        return String.format("%slet payload = payloads[Math.floor(Math.random() * payloads.length)];%s",
                newLine, newLine);
    }

    private String randomQueryParamScript() {
        return """
                let currentSearchParams = searchParams[Math.floor(Math.random() * searchParams.length)];
                let urlSearchParams = new URLSearchParams(currentSearchParams);
                """;
    }
}

File: src/main/kotlin/dqualizer/dqexec/loadtest/mapper/k6/HttpMapper.kt
package dqualizer.dqexec.loadtest.mapper.k6

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import dqualizer.dqexec.config.ResourcePaths
import dqualizer.dqexec.exception.UnknownRequestTypeException
import io.github.dqualizer.dqlang.types.adapter.request.Request
import java.util.*
import org.springframework.stereotype.Component

/** Maps the specified request to a k6 'default function()' */
@Component
class HttpMapper(private val resourcePaths: ResourcePaths) : K6Mapper {
  override fun map(request: Request): String {
    val httpBuilder = StringBuilder()
    httpBuilder.append(exportFunctionScript())
    val path = request.path
    val type = request.type.uppercase(Locale.getDefault())
    val method =
      when (type) {
        "GET" -> "get"
        "POST" -> "post"
        "PUT" -> "put"
        "DELETE" -> "del"
        else -> throw UnknownRequestTypeException(type)
      }

    // Code for choosing one random variable for every existing path variable
    val pathVariables = request.pathVariables
    val maybeReference = pathVariables.stream().findFirst()
    if (maybeReference.isPresent) {
      val pathVariablesString =
        resourcePaths.readResourceFile(maybeReference.get().scenarios[0].path)
      try {
        val node = ObjectMapper().readTree(pathVariablesString)
        val variables = node.fieldNames()
        while (variables.hasNext()) {
          val variable = variables.next()
          val randomPathVariable = randomPathVaribleScript(variable)
          httpBuilder.append(randomPathVariable)
        }
      } catch (e: JsonProcessingException) {
        e.printStackTrace()
      }
    }

    // Code for using the request-parameter and payload inside the http method
    val payload = request.payload
    val queryParams = request.queryParams
    var extraParams = ""
    if (!payload.isEmpty() || !queryParams.isEmpty()) {
      extraParams =
        if (!payload.isEmpty() && !queryParams.isEmpty()) {
          httpBuilder.append(randomQueryParamScript())
          httpBuilder.append(randomPayloadScript())
          " + `?\${urlSearchParams.toString()}`, JSON.stringify(payload)"
        } else if (!payload.isEmpty()) {
          httpBuilder.append(randomPayloadScript())
          ", JSON.stringify(payload)"
        } else {
          httpBuilder.append(randomQueryParamScript())
          " + `?\${urlSearchParams.toString()}`"
        }
    }
    val httpRequest =
      java.lang.String.format(
        "%slet response = http.%s(baseURL + `%s`%s, params);%s",
        System.lineSeparator(),
        method,
        path,
        extraParams,
        System.lineSeparator(),
      )
    httpBuilder.append(httpRequest)
    return httpBuilder.toString()
  }

  private fun exportFunctionScript(): String {
    return String.format(
      "%sexport default function() {%s",
      System.lineSeparator(),
      System.lineSeparator(),
    )
  }

  private fun randomPathVaribleScript(variable: String): String {
    return String.format(
      "%slet %s = %s_array[Math.floor(Math.random() * %s_array.length)];%s",
      System.lineSeparator(),
      variable,
      variable,
      variable,
      System.lineSeparator(),
    )
  }

  private fun randomPayloadScript(): String {
    return String.format(
      "%slet payload = payloads[Math.floor(Math.random() * payloads.length)];%s",
      System.lineSeparator(),
      System.lineSeparator(),
    )
  }

  private fun randomQueryParamScript(): String {
    return """
            let currentSearchParams = searchParams[Math.floor(Math.random() * searchParams.length)];
            let urlSearchParams = new URLSearchParams(currentSearchParams);
    """
      .trimIndent()
  }
}
